src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -84,43 +94,144 @@ private Task OnProjectFactoryCompletedAsync()
+        {
+            lock (_projectContextProvider)
+            {
+                if (string.Equals(targetFrameworks, _latestTargetFrameworks, StringComparison.OrdinalIgnoreCase))
+                {
+                    // We have already handled this targetFrameworks update.
+                    return;
+                }
+            }
+
+            var newProjectContext = await _projectContextProvider.UpdateProjectContextAsync(targetFrameworks, CancellationToken.None).ConfigureAwait(false);
+
+            AggregateWorkspaceProjectContext previousProjectContext;
+            lock (_projectContextProvider)
+            {
+                if (newProjectContext == _aggregateProjectContext)
+                {
+                    // Another thread has already completed the update.
+                    Requires.Range(string.Equals(targetFrameworks, _latestTargetFrameworks, StringComparison.OrdinalIgnoreCase), nameof(targetFrameworks));
+                    return;
+                }
+
+                previousProjectContext = _aggregateProjectContext;
+                _aggregateProjectContext = newProjectContext;
+                _latestTargetFrameworks = targetFrameworks;
+            }
+
+            await ResetSubscriptionsAsync().ConfigureAwait(false);
+
+            if (previousProjectContext != null)
+            {
+                foreach (var innerContext in previousProjectContext.InnerProjectContexts)
+                {
+                    foreach (var handler in Handlers)
+                    {
+                        await handler.Value.OnContextReleasedAsync(innerContext).ConfigureAwait(false);
+                    }
+                }
             }
         }
 
-        private async Task HandleAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e, RuleHandlerType handlerType)
+        private async Task ResetSubscriptionsAsync()
+        {
+            DisposeAndClearSubscriptions();
+
+            using (_tasksService.LoadedProject())
+            {
+                var watchedEvaluationRules = GetWatchedRules(RuleHandlerType.Evaluation);
+                var watchedDesignTimeBuildRules = GetWatchedRules(RuleHandlerType.DesignTimeBuild);
+                var activeConfiguredProjects = await _activeConfiguredProjectsProvider.GetActiveConfiguredProjectsAsync().ConfigureAwait(false);
+
+                foreach (var project in activeConfiguredProjects)
+                {
+                    _designTimeBuildSubscriptionLinks.Add(_activeConfiguredProjectSubscriptionService.JointRuleSource.SourceBlock.LinkTo(
+                        new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(e => OnProjectChangedCoreAsync(e, RuleHandlerType.DesignTimeBuild)),
+                        ruleNames: watchedDesignTimeBuildRules.Union(watchedEvaluationRules), suppressVersionOnlyUpdates: true));
+
+                    _evaluationSubscriptionLinks.Add(project.Services.ProjectSubscription.ProjectRuleSource.SourceBlock.LinkTo(
```

Comment:
>project.Services.ProjectSubscription.ProjectRuleSource.SourceBlock.LinkTo [](start = 53, length = 73)

Since you are listening separately to each configured project (1) you get a different call to OnProjectChangedCoreAsync for each tfm - this appears to be OK in your scenario, and (2) ProjectPropertyChangedAsync below is no longer needed